src/api/syscall.c

```diff
@@ -317,7 +321,12 @@ exception_t handleUnknownSyscall(word_t w)
              * value. For many architectures, the TLS variables only get
              * updated on a thread switch.
              */
-            return Arch_setTLSRegister(tls_base);
+            exception_t ret = Arch_setTLSRegister(tls_base);
+            if (unlikely(ret != EXCEPTION_NONE)) {
+                userError("could not set TLS register");
```

Comment:
I can do this also - but then why don't we let this return `bool_t` at least to have a way to report an error up to the generic code. Seem there is a code path that may have an error and for debug/dev builds it can't be bad to see some messages.

Reply:
Ok, I looked at the bigger context of that function, and the whole `handleUnknownSyscall` is a bit of a disaster area, fusing different parts of the protocol together. I guess the idea is that none of these are for verification, so structure is less important. At least for `Arch_setTLSRegister` the code path with error does seem Ok in the sense that no state change happens before the error is raised, and `exception_t` is the right one to use here, even though there's no distinction which phase what is in.

So, let's use most of your proposal, but in the generic code we should check for `EXCEPTION_SYSCALL_ERROR` when we report an error and pass up any other `exception_t` values we might get.

Reply:
After having looked at the corresponding Haskell type for `handleUnknownSyscall`, you are completely right: the function has to be `void`, or if it returns `exception_t`, it is only allowed to return `EXCEPTION_NONE`. It's not inside a preemption handler, so it cannot return preemption exceptions either (and it cannot return failures or other results).

So the existing version is definitely wrong, and your suggested change is actually fully correct.

Reply:
Besides the `userError()`, should this also add a `halt()` in case of error? This would change the behavior, but it might make more sense than hiding the error. It's a dev feature that is not supposed to fail and not returning a result, so halting might be the best option to have a sane behavior then.

Reply:
No, user behaviour should never be able to trigger a `halt`. The potential return value of an execute function after decode can only indicate preemption or not, error is no longer an option, so this code is definitely wrong in any case (i.e. checking for return value and reporting an error here). 

All user errors have to be caught in the decode phase before any state change happens.